crates/swift-bridge-ir/src/codegen/codegen_tests/option_codegen_tests.rs

```diff
@@ -460,6 +460,122 @@ void* __swift_bridge__$some_function(void);
+mod extern_rust_fn_return_option_ref_opaque_rust_type {
+    use super::*;
+
+    fn bridge_module_tokens() -> TokenStream {
+        quote! {
+            mod ffi {
+                extern "Rust" {
+                    type SomeType;
+                    fn some_function () -> Option<&SomeType>;
+                }
+            }
+        }
+    }
+
+    fn expected_rust_tokens() -> ExpectedRustTokens {
+        ExpectedRustTokens::Contains(quote! {
+            #[export_name = "__swift_bridge__$some_function"]
+            pub extern "C" fn __swift_bridge__some_function() -> *const super::SomeType {
+                if let Some(val) = super::some_function() {
+                    val as *const super::SomeType
+                } else {
+                    std::ptr::null()
+                }
+            }
+        })
+    }
+
+    fn expected_swift_code() -> ExpectedSwiftCode {
+        ExpectedSwiftCode::ContainsAfterTrim(
+            r#"
+func some_function() -> Optional<SomeTypeRef> {
+    { let val = __swift_bridge__$some_function(); if val != nil { return SomeTypeRef(ptr: val!) } else { return nil } }()
+}
+"#,
+        )
+    }
+
+    fn expected_c_header() -> ExpectedCHeader {
+        ExpectedCHeader::ContainsAfterTrim(
+            r#"
+void* __swift_bridge__$some_function(void);
+    "#,
+        )
+    }
+
+    #[test]
+    fn extern_rust_fn_return_option_opaque_rust_type() {
+        CodegenTest {
+            bridge_module: bridge_module_tokens().into(),
+            expected_rust_tokens: expected_rust_tokens(),
+            expected_swift_code: expected_swift_code(),
+            expected_c_header: expected_c_header(),
+        }
+        .test();
+    }
+}
+
+/// Test code generation for Rust function that returns an Option<&OpaqueRustType>
+mod extern_rust_fn_arg_option_ref_opaque_rust_type {
+    use super::*;
+
+    fn bridge_module_tokens() -> TokenStream {
+        quote! {
+            mod ffi {
+                extern "Rust" {
+                    type SomeType;
+                    fn some_function (arg: Option<&SomeType>) -> Option<&SomeType>;
+                }
+            }
+        }
+    }
+
+    fn expected_rust_tokens() -> ExpectedRustTokens {
+        ExpectedRustTokens::Contains(quote! {
+            #[export_name = "__swift_bridge__$some_function"]
+            pub extern "C" fn __swift_bridge__some_function() -> *const super::SomeType {
+                if let Some(val) = super::some_function() {
+                    val as *const super::SomeType
+                } else {
+                    std::ptr::null()
+                }
+            }
+        })
+    }
+
+    fn expected_swift_code() -> ExpectedSwiftCode {
+        ExpectedSwiftCode::ContainsAfterTrim(
+            r#"
+func some_function(arg: Optional<SomeTypeRef>) -> Optional<SomeTypeRef> {
+    { let val = __swift_bridge__$some_function(); if val != nil { return SomeTypeRef(ptr: val!) } else { return nil } }()
+}
```

Comment:
This one is similar to: https://github.com/chinedufn/swift-bridge/blob/607e30579ac4de1fab5ff66838253cf3d07ddfbb/crates/swift-bridge-ir/src/codegen/codegen_tests/extern_rust_function_opaque_rust_type_argument_codegen_tests.rs#L91-L99

So it can be:

```swift
func some_function(_ arg: Optional<SomeTypeRef>) {
    __swift_bridge__$some_function({ if let val = arg { return val.ptr } else { return nil } }())
}
```

Reply:
Seems that generated is this:

```swift
func some_function(_ arg: Optional<SomeTypeRef>) {
    __swift_bridge__$some_function({ if let val = arg { val.isOwned = false; return val.ptr } else { return nil } }())
}

```

Just checking that the assignment to setting isOwned to false is correct?

Reply:
It does by the way look like codegen on the previous test (that you told me to change to `return SomeTypeRef>` is actually `return SomeType` - I'm guessing this is a mistake and something I need to correct in translation to swift somewhere. Do you know where that would be?

Reply:
`isOwned` is available on owned `SomeType` https://github.com/chinedufn/swift-bridge/blob/dd5bef56af28db4f1a1244d86115def6abede931/crates/swift-bridge-ir/src/codegen/codegen_tests/opaque_rust_type_codegen_tests.rs#L36-L48

but not on `SomeTypeRef` https://github.com/chinedufn/swift-bridge/blob/dd5bef56af28db4f1a1244d86115def6abede931/crates/swift-bridge-ir/src/codegen/codegen_tests/opaque_rust_type_codegen_tests.rs#L54-L60

---

So, we just need what should hopefully be one last `if self.reference` here https://github.com/chinedufn/swift-bridge/blob/4fbd30f81085dc366bda3c0303eac66345de3ed9/crates/swift-bridge-ir/src/bridged_type/bridged_opaque_type.rs#L312-L325

Reply:
Oh wait - that type has access to know if it's a reference - in that case I can just modify `swift_name` to check `if self.reference` and append the suffix accordingly

Reply:
Hahaha look what I was about to comment with before I saw these last two comments:

https://github.com/chinedufn/swift-bridge/blob/4fbd30f81085dc366bda3c0303eac66345de3ed9/crates/swift-bridge-ir/src/bridged_type/bridged_opaque_type.rs#L639-L642

```rust
pub fn swift_name(&self) -> String {
	if self.reference {
		format!("{}", self.ty)
	} else {
		format!("{}Ref", self.ty)
	}
}
```

Your FFI-fu has grown strong my friend.

Reply:
I think that as you start handling more cases the two languages start to diverge but based on what we currently have tested they're always the same.

As in, you end up starting to add certain if statements based on the language.

So what happened here was probably something like:

1. if rust do X, if swift `todo!()`
2. one day we needed to add support for Swift so we wrote a test case
3. then we changed the logic to -> if rust do X, if swift do X
4. FUTURE -> if rust do X or Y depending on Z, if swift do X or Q depending on W

So, this is just in the middle of that transition to step 4.

I suppose you could collapse then if you wanted, but I think it's better as is so that when someone someday needs to modify one language they can clearly see where to do it (I haven't had to do this in a while so I forget what can cause the divergence).

If you disagree and think we should collapse back down until that day comes, feel free to collapse it.

Reply:
One more spot I needed to check reference, but all working now!

In `convert_option_swift_expression_to_ffi_type`